### ISSUE TITLE
How to overlap subscriber view over publisher view

### ISSUE THREAD

**Issue by raulgrecio** [opentok/opentok-react-native#162]
I'm using v0.8.4
the controls and the subscriber are renderer under the publisher.

This is the opposite of issue #122

```
render() {
    const { height, width } = Dimensions.get('window')

    return (
      <View style={{ flex: 1 }}>
        <OTSession apiKey={this.apiKey} sessionId={this.sessionId} token={this.token}>
          <OTSubscriber style={{ position: 'absolute',  top: 20, right: 20, width: 100, height: 100, zIndex: 1000 }}
          />
          <OTPublisher style={{ width, height }} />
        </OTSession>
        <View
          style={{ position: 'absolute', bottom: 20, left: 0, right: 0, zIndex: 1000, alignItems: 'center', justifyContent: 'center'}}
        >
          <View
            // simulates controls
            style={{
              backgroundColor: 'red',
              width: 60,
              height: 60,
              borderRadius: 30,
            }}
          />
        </View>
      </View>
    )
}
```

**Comment by J-A-B-R** [comment]
We are facing the same proble with 0.8.4

We are developing a streaming app, so our use case is that the publisher should fill the screen and the subscriber/s overlap on top of it.

In iOS there is no problem, but in Android the publisher is always above the subscriber or any other component no matter what we try like putting OTSubscriber before OTPublisher, playing with *absolute* and *zIndex* or other weird and twisted code.

We think that the problem lies on the change in commit c5d979f1f8a6d775148083ea6105e47b936adf70 that forces this behavior with `((GLSurfaceView) mPublisher.getView()).setZOrderOnTop(true)`.

We cannot revert to 0.8.3 because the current version also fixes issue #145 so, could be possible in the next patch to add a prop to OTPublisher and OTSubscriber to control if the component will be on top or not?

**Comment by msach22** [comment]
@J-A-B-R Sorry about breaking your app with the latest release 😞. I will see what I can do and propose a new API, something like onTop, or see if we can use zIndex value of the style prop to determine which stays on top.

**Comment by J-A-B-R** [comment]
@msach22 I was suggesting *onTop* as a simple condition to choose to execute or not `setZOrderOnTop(true)`. But if you can implement it with *zIndex*, it would be great and even more consistent with the rest of RN :ok_hand:.

**Comment by akandaurov** [comment]
Same issue for me as well - Publisher video is over everything. Worked fine before 0.8.4

**Comment by J-A-B-R** [comment]
I've been digging a little into this problem in this slow Sunday and I have two suggestions for the z-index problem:

1. Use *TextureView* instead of *SurfaceView* as [supported by the OpenTok SDK](https://tokbox.com/developer/sdks/android/reference/com/opentok/android/Session.SessionOptions.html "SessionOptions"). The only caveats are, as far as I know, that the window must be hardware accelerated and it's expected a latency of one or two FPS.

2. But if you stick with the *SurfaceView* route, then consider using *setZOrderMediaOverlay* instead of *setZOrderOnTop*. This way, the on top surface would be still below the UI window ([Google grafika example](https://github.com/google/grafika/blob/master/app/src/main/java/com/android/grafika/MultiSurfaceActivity.java "MultiSurfaceActivity")).

Hope this helps.

**Comment by sunweiyang** [comment]
@msach22 Will this issue be addressed soon? Publisher is always rendered over everything else on Android, which makes overlays or overlaps on top of the publisher impossible.

**Comment by msach22** [comment]
@sunweiyang Yes, I will be checking this out this week

**Comment by msach22** [comment]
@sunweiyang  @J-A-B-R @akandaurov I've been investigating this and have a couple of questions for the community. 

- If we were to use a prop such as `onTop`, what should happen when both `OTPublisher` and `OTSubscriber` are supplied with `onTop`? 
- Should we only check the first one?

 I've tried to do use the `zIndex ` value and passing it in the and `setZ` and `setTranslationZ` methods, but that isn't doing the trick. I'm also updating all of the views when a new publisher or subscriber are created, but this didn't work. The only thing that's allowing me to add the view on top is if I do: `setZOrderOnTop(true)`.

Let me know what you folks think. Is it better to completely remove the `setZOrderOnTop(true)` for the publisher while we attempt to resolve this? I can release a patch for that ASAP so it doesn't mess with your current work.

**Comment by J-A-B-R** [comment]
@msach22 Hi and happy new year, as I understand it, a **SurfaceView** is not a real UI view. Instead of that, all the true views on the UI are rendered in a **SurfaceView** created by default by Android. Any other **SurfaceViews** you create are independent of this one. The only method to control the Z index of **SurfaceViews** is with `setZOrderOnTop` and `setZOrderMediaOverlay`. So there is a sort of four Z levels for **SurfaceViews**:
 
0. Any **SurfaceView** you create.
1. Any **SurfaceView** you create and on which you call the method `setZOrderMediaOverlay` (good if you want, for example, a control overlay over a video).
2. The UI **SurfaceView** (the regular view hierarchy goes here).
3. Any **SurfaceView** you create and on which you call the method `setZOrderOnTop`.

However, I don't know how Android sorts `SurfaceViews` on the same level.

Another possibility is to use a **TextureView** that is a proper view on the UI and behaves like any other view. The trade-off is that by its very nature is less performant than a **SurfaceView**.

For anyone who wants to deepen in the gore details: https://source.android.com/devices/graphics

As for you question, in my humble opinion:

- Leave as default behavior whatever one that represents the least disruption to the current user base. In our case, we had to choose to not overlap the surfaces. Our client was satisfied with this solution (our layout guy did a great job here), so we are happy with any default behavior yo choose).
- Define the prop to be of type `'none' | 'media' | 'top'` for hinting what `setZorder...` method to apply or not.
- It's up to the developer how he uses these props as long as he is well aware of how Android sorts the surfaces as explained above.

Hope this helps.

**Comment by xxsnakerxx** [comment]
Using `useTextureViews` works for me!!! Also fixes issues with border radius https://github.com/opentok/opentok-react-native/issues/174

https://github.com/opentok/opentok-android-sdk-samples/issues/136#issuecomment-322218524 

```java
@ReactMethod
    public void initSession(String apiKey, String sessionId) {

        Session mSession = new Session.Builder(this.getReactApplicationContext(), apiKey, sessionId)
            // add session options
            .sessionOptions(new Session.SessionOptions() {
                @Override
                public boolean useTextureViews() {
                    return true;
                }
            })
            .build();
        mSession.setSessionListener(this);
        mSession.setSignalListener(this);
        mSession.setConnectionListener(this);
        mSession.setReconnectionListener(this);
        mSession.setArchiveListener(this);
        mSession.setStreamPropertiesListener(this);
        sharedState.setSession(mSession);
    }
```

**Comment by msach22** [comment]
@xxsnakerxx Does this resolve the zIndex issue for you?

@J-A-B-R Thanks for your investigation! We can expose the `TextureView` option via the `properties` prop in `OTSession` component which would then allow the application to be have they way they want. What do you think?

**Comment by xxsnakerxx** [comment]
@msach22 Yes, resolved, completely

**Comment by msach22** [comment]
@xxsnakerxx Were you using the following code on the JS end to test?
https://github.com/opentok/opentok-react-native/issues/162#issue-358183035

**Comment by xxsnakerxx** [comment]
@msach22 No, but in my project, the code looks similar and without this fix fullscreen publisher did overlap subscribers.

**Comment by msach22** [comment]
@xxsnakerxx Unfortunately, this doesn't seem the resolve the issue that I'm having. I will keep you updated on my findings

**Comment by akandaurov** [comment]
Just roll back the fix you did for #122 and everything works fine. In order to put the publisher over subcriber before that fix it just should be wrapped in additional view with absolute position. I had to make my fork and roll back that fix because it broke everything for us - there is no way to put anything over the publisher.
I also had to make fixes to 100500 crashes on Android in my fork - but that's another story.

**Comment by J-A-B-R** [comment]
@msach22 Exposing through `OTSession`'s `properties` prop seems like an excellent idea. To get the best of both worlds, it could be somthing like: `androidView: 'textureView' | 'surfaceViewPublisherOnTop' | 'surfaceViewPublisherOnMediaOverlay' | 'surfaceViewSubscriberOnTop' | 'surfaceViewSubscriberOnMediaOverlay'`. Verbose, I know, but it could state clearly the code intention.

At least, until you resolve any other pending issues, it will unbreak current projects whatever its use case is.

**Comment by evillemez** [comment]
Is this addressed in the current `0.9.7` version by any chance?

I think we're running into similar issues where we want to switch which feed is full screen depending on the situation, but our UI isn't overlaying over the publisher, so no luck yet.

**Comment by bob76828** [comment]
Same issue as @evillemez. 
Does anyone can help us ?

**Comment by Auggysg** [comment]
Hi I'm facing similar issue where I'm trying to add a full screen text component while in the call room but I can't overlap the publisher. I'm not sure where to find the TextureView option or the setZOrderOnTop options. 

Any help would be much appreciated! Been trying to figure this out for a couple of days...

**Comment by bob76828** [comment]
@Auggysg Hope This is helpful for you.
https://github.com/mediot-inc/opentok-react-native/commit/68a3e64af491d2ef5318a64bdd0b225db99fc1e0

**Comment by msach22** [comment]
We've removed the forcing of the zIndex for the publisher on Android. I'll be working with @ggoldens on this and update you as we make some progress.

**Comment by msach22** [comment]
This has been released with `v0.10.0`. Thank you all for contributing and helping define the fix. Please check out the docs to see how you can use it: https://github.com/opentok/opentok-react-native/blob/master/docs/OTSession.md#setting-session-options

I'll also work on a sample application for this.

**Comment by HayTran94** [comment]
> @msach22 Hi and happy new year, as I understand it, a **SurfaceView** is not a real UI view. Instead of that, all the true views on the UI are rendered in a **SurfaceView** created by default by Android. Any other **SurfaceViews** you create are independent of this one. The only method to control the Z index of **SurfaceViews** is with `setZOrderOnTop` and `setZOrderMediaOverlay`. So there is a sort of four Z levels for **SurfaceViews**:
> 
> 1. Any **SurfaceView** you create.
> 2. Any **SurfaceView** you create and on which you call the method `setZOrderMediaOverlay` (good if you want, for example, a control overlay over a video).
> 3. The UI **SurfaceView** (the regular view hierarchy goes here).
> 4. Any **SurfaceView** you create and on which you call the method `setZOrderOnTop`.
> 
> However, I don't know how Android sorts `SurfaceViews` on the same level.
> 
> Another possibility is to use a **TextureView** that is a proper view on the UI and behaves like any other view. The trade-off is that by its very nature is less performant than a **SurfaceView**.
> 
> For anyone who wants to deepen in the gore details: https://source.android.com/devices/graphics
> 
> As for your question, in my humble opinion:
> 
> * Leave as default behavior whatever one that represents the least disruption to the current user base. In our case, we had to choose to not overlap the surfaces. Our client was satisfied with this solution (our layout guy did a great job here), so we are happy with any default behavior yo choose).
> * Define the prop to be of type `'none' | 'media' | 'top'` for hinting what `setZorder...` method to apply or not.
> * It's up to the developer how he uses these props as long as he is well aware of how Android sorts the surfaces as explained above.
> 
> Hope this helps.

Thank you very much. According your knowledge about surfaceView Z index level, I can implement my use case in conference room where having many subscriber and one publisher. The publisher and minor subscriber is above major subscriber. Once again. Thank you a lots.
